File: 1048-longest-string-chain/1048-longest-string-chain.java
class Solution {
    public int longestStrChain(String[] words) {
        Set<String> set = new HashSet<>();
        for(int i =0;i<words.length;i++){
            set.add(words[i]);
        }
        Map<String,Integer> map = new HashMap<>();
        int res = 0;
        for(int i =0;i<words.length;i++){
            res = Math.max(dfs(words[i],set,map),res);
        }
        return res;
    }
    
    
    public int dfs(String s, Set<String> set, Map<String,Integer> map){
        if(!set.contains(s)){
            return 0;
        }
        if(s.length()==1){
            return 1;
        }
        if(map.containsKey(s)){
            return map.get(s);
        }
        int tempRes = 0;
        for(int i=0;i<s.length();i++){
            String temp = s.substring(0,i)+s.substring(i+1,s.length());
            tempRes = Math.max(tempRes,1+dfs(temp,set,map));
        }
        map.put(s,tempRes);
        return tempRes;
    }
}

File: 1048-longest-string-chain/1048-longest-string-chain.py
class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        
        def isPred(s1,s2):
            p1 = len(s1)-1
            p2 = len(s2)-1
            oneViolation = False
            if(p2-p1!=1):
                return False
            while(p1>=0 or p2>=0):
                if(p1>=0 and s2[p2]==s1[p1]):
                    p1=p1-1
                    p2=p2-1
                else:
                    if(oneViolation):
                        return False
                    p2=p2-1
                    oneViolation = True
            if(p1==-1 and p2==-1):
                return True
            return False
            
        words = sorted(words,key=len)
        n = len(words)
        dp = [1]*n
        ans = 1
        for i in range(1,n):
            for j in range(i):
                if(isPred(words[j],words[i]) and dp[i]<dp[j]+1):
                    dp[i] = dp[j] + 1
            ans = max(ans,dp[i])
            
    
        return ans
        
        
        
        
                
            
            
                     
                     
                     
                                 
            
                     